codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,154 @@
+                                                                    Unpooled.wrappedBuffer(new byte[] {'\r', '\n'})));
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, RedisMessage msg, List<Object> out) throws Exception {
+        try {
+            writeRedisMessage(ctx.alloc(), msg, out);
+        } catch (CodecException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new CodecException(e);
+        }
+    }
+
+    private static void writeRedisMessage(ByteBufAllocator allocator, RedisMessage msg, List<Object> out) {
+        if (msg instanceof StringRedisMessage) {
+            writeInlineByteBufMessage(allocator, (StringRedisMessage) msg, out);
+        } else if (msg instanceof IntegerRedisMessage) {
+            writeIntegerMessage(allocator, (IntegerRedisMessage) msg, out);
+        } else if (msg instanceof FullBulkStringRedisMessage) {
+            writeFullBulkStringMessage(allocator, (FullBulkStringRedisMessage) msg, out);
+        } else if (msg instanceof BulkStringRedisContent) {
+            writeBulkStringContent((BulkStringRedisContent) msg, out);
+        } else if (msg instanceof BulkStringHeaderRedisMessage) {
+            writeBulkStringHeader(allocator, (BulkStringHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayHeaderRedisMessage) {
+            writeArrayHeader(allocator, (ArrayHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayRedisMessage) {
+            writeArrayMessage(allocator, (ArrayRedisMessage) msg, out);
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+    }
+
+    private static void writeInlineByteBufMessage(ByteBufAllocator allocator, StringRedisMessage msg,
+                                                  List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        out.add(ByteBufUtil.encodeString(allocator, CharBuffer.wrap(msg.content()), CharsetUtil.UTF_8));
+        out.add(EOL.slice());
+    }
+
+    private static void writeIntegerMessage(ByteBufAllocator allocator, IntegerRedisMessage msg, List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        out.add(numberToByteBuf(msg.value()));
+        out.add(EOL.slice());
+    }
+
+    private static void writeBulkStringHeader(ByteBufAllocator allocator, BulkStringHeaderRedisMessage msg,
+                                              List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            out.add(numberToByteBuf(msg.bulkStringLength()));
+            out.add(EOL.slice());
+        }
+    }
+
+    private static void writeBulkStringContent(BulkStringRedisContent msg, List<Object> out) {
+        out.add(msg.content().retain());
```

Comment:
I don't think we should retain on the encode side. If someone writes an object they assume reference count ownership is transferred.